### DISCUSSION TITLE
Backwards compatible `.pwsh` extension

### DISCUSSION THREAD

**Discussion by Stan-Stani** [PowerShell/PowerShell#24112] [Polls]
I wish we could use a .pwsh extension to run with pwsh core and provide a warning if it tries to execute with powershell (because that's all that's on the system) for windows but still try to execute it...

**Comment by ThomasNieto** [comment]
I'd rather script and module authors verify what versions their scripts run on and then put that in the manifest so there is no ambiguity. A warning and running the script anyway just creates user confusion especially when it partially fails.

**Reply by hrxn** [reply]
Why not do both?

**Comment by rhubarb-geek-nz** [comment]
Automatic mappings for console programs are really bad when they just start and close down and you never see what they actually did.

If you personally want a pwsh mapping for your system you are welcome to set one up, but I certainly would not recommend it as part of the standard product.

**Comment by kasini3000** [comment]
Attention: Immature, for discussion only
I have some suggestions here:

1. Add an extension. ps2
2. Continuously improve PS2 without affecting PS1 at all:
Organize a list and discuss the features that should be deprecated
ps2 is only compatible with PowerShell v5.1 and PowerShell v7.6
Abandon those old syntax features that cannot be changed for PS1 compatibility.
3. Define a ps2config.xml. Define a command that allows switching on and off each feature.

**Reply by rhubarb-geek-nz** [reply]
>    ps2 is only compatible with PowerShell v5.1 and PowerShell v7.6

It is worth remembering that PowerShell v5.1 is effectively end-of-life/maintenance mode and not related to this PowerShell v7 project.

**Comment by hrxn** [comment]
In favor, I think this would be a great idea.

**Comment by rhubarb-geek-nz** [comment]
> but still try to execute it...

Who is doing the "but still try to execute it" ?

Out of the box on Windows, there is no extension mapping for .ps1 files to PowerShell. Attempting to run from either cmd.exe or Windows Explorer presents you with a dialog asking you to select a program to run.

Likewise there is no mapping for a file with the .pwsh extension. So from that point of view .ps1 and .pwsh behave exactly the same.

If you run

```
powershell .\foo.pwsh
```

You still get the same dialog.

It you did set up the extensions mapping, PowerShell 5.1 itself would not know what to do with the file, it handles .ps1 files and delegates other file types to windows. So that is what it will do.

Who are you expecting to provide the extension mapping and how will it make PowerShell 5.1 treat it as a script?

We already have an extension that is supported by both PowerShell and PowerShell Core, the `.ps1`. If a script wants to print a warning or throw an error if run by PowerShell 5.1 it can already do so.

**Comment by kilasuit** [comment]
I'll add some history here

There was the option for a ps2 extension so that in PowerShell v2 if desired we could have enabled per version script versioning ( PowerShell v2 could run both ps1 and ps2 files (as well as all other variants that PowerShell supports like psd1, psm1 etc) whereas PowerShell v1 could only run those files with a 1 in it) 

This was dropped for the [#Requires -Version directive](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-7.4#-version-nn) which essentially means that this request is already coded for in a more flexible model than using per version file associations & is IMO not going to change now.

If you want to manage scripts that can only run in PowerShell version 7+ then you can add code like `#Requires -Version 7` to stop it working in v5 or code to automatically re-run it in pwsh as opposed to powershell but the work to add support for a `.pwsh` extension is not a small amount of work and like .ps2 never happened, I would be willing to say that this just isn't going to happen for any future version.

**Comment by kasini3000** [comment]
> PowerShell v2 could run both ps1 and ps2 files 

Place 2 engines in powershell.exe:
All. ps1 run with old engine
All new engines executing. ps2 and ps2 are destructive updates that need to be eliminated and removed from issues that cannot be changed for compatibility

**Reply by rhubarb-geek-nz** [reply]
> Place 2 engines in powershell.exe:

I don't see the motivation for this. Both PowerShell.exe and pwsh.exe run ps1 files. If there is any major compatibility break it should be about not supporting PowerShell.exe and requiring pwsh.exe which can already be done with the "Requires" feature as described above. PowerShell.exe is on end-of-life/life-support and not related to this project. I don't see the point in trying to add so called 'ps2 features' to PowerShell.exe.
